java/memory/src/main/java/org/apache/arrow/memory/util/ArrowBufPointer.java

```diff
@@ -1,106 +1,147 @@
+ */
+
+package org.apache.arrow.memory.util;
+
+import org.apache.arrow.memory.util.hash.ArrowBufHasher;
+import org.apache.arrow.memory.util.hash.DirectHasher;
+import org.apache.arrow.util.Preconditions;
+
+import io.netty.buffer.ArrowBuf;
+
+/**
+ * Pointer to a memory region within an {@link io.netty.buffer.ArrowBuf}.
+ * It will be used as the basis for calculating hash code within a vector, and equality determination.
+ */
+public final class ArrowBufPointer {
+
+  /**
+   * The hash code when the arrow buffer is null.
+   */
+  public static final int NULL_HASH_CODE = 0;
+
+  private ArrowBuf buf;
+
+  private int offset;
+
+  private int length;
+
+  private int hashCode = NULL_HASH_CODE;
+
+  private ArrowBufHasher hasher = DirectHasher.INSTANCE;
+
+  /**
+   * A flag indicating if the underlying memory region has changed.
+   */
+  private boolean hashCodeChanged = false;
+
+  /**
+   * The default constructor.
+   */
+  public ArrowBufPointer() {
+
+  }
+
+  /**
+   * Constructs an Arrow buffer pointer.
+   * @param buf the underlying {@link ArrowBuf}, which can be null.
+   * @param offset the start off set of the memory region pointed to.
+   * @param length the length off set of the memory region pointed to.
+   */
+  public ArrowBufPointer(ArrowBuf buf, int offset, int length) {
+    set(buf, offset, length);
+  }
+
+  /**
+   * Sets this pointer.
+   * @param buf the underlying {@link ArrowBuf}, which can be null.
+   * @param offset the start off set of the memory region pointed to.
+   * @param length the length off set of the memory region pointed to.
+   */
+  public void set(ArrowBuf buf, int offset, int length) {
+    this.buf = buf;
+    this.offset = offset;
+    this.length = length;
+
+    hashCodeChanged = true;
+  }
+
+  /**
+   * Sets the hasher for the pointer.
+   * @param hasher the hasher to set.
+   */
+  public void setHasher(ArrowBufHasher hasher) {
+    Preconditions.checkNotNull(hasher);
+    this.hasher = hasher;
+    hashCodeChanged = true;
+  }
+
+  /**
+   * Gets the underlying buffer, or null if the underlying data is invalid or null.
+   * @return the underlying buffer, if any, or null if the underlying data is invalid or null.
+   */
+  public ArrowBuf getBuf() {
+    return buf;
+  }
+
+  public int getOffset() {
+    return offset;
+  }
+
+  public int getLength() {
+    return length;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+
+    ArrowBufPointer other = (ArrowBufPointer) o;
+    if (buf == null || other.buf == null) {
+      if (buf == null && other.buf == null) {
+        return true;
+      } else {
+        return false;
+      }
+    }
+
+    return ByteFunctionHelpers.equal(buf, offset, offset + length,
+            other.buf, other.offset, other.offset + other.length) != 0;
+  }
+
+  @Override
+  public int hashCode() {

```

Comment:
this seems to provide for the posibility  that hashCodes are not equals when when two pointers are equal.  I don't have a great solution here, just a comment.

Reply:
Do you mean the cases where the data content changes, while the address and length remain the same?

Reply:
@emkornfield can you please explain when this possibility occurs ?

Reply:
I was thinking of two different hash functions in use?

Reply:
would it be cleaner if the hasher can be set only via the constructor ? That way, a given instance of ArrowBufPointer cannot return different hash values for the same data content,

Reply:
@pravindra Great suggestion. Thank you so much.

Reply:
this lgtm. @emkornfield does this solve the concern you raised ?

Reply:
I don't want to hold up merging.  You still have a strange contract that two separately constructed pointers can have different hashcodes but are still equals.  One could make equality date into account that hasher but that might be slightly surprising as well.

Reply:
you are right - this will get tricky if ArrowBufPointer are used as keys in a HashMap. I think using the hasher as part of the equality check is the cleanest way to avoid bugs. I think it solves all of the conditions in the contract

https://docs.oracle.com/javase/7/docs/api/java/lang/Object.html#hashCode%28%29

Reply:
@pravindra Good point. 
Incorporating the hasher into equality check essentially incorporates the Hasher type into the type system of ArrowBufPointer